File: algorithms/hashing/anagrams/anagrams.js
// Helper Function
const charFreqHM = (word) => {
  let freqHashMap = {};

  for (let char of word) {
    if (!(char in freqHashMap)) {
      freqHashMap[char] = 0;
    }

    freqHashMap[char] += 1;
  }

  return freqHashMap;
};

// Main Driver Function
const anagrams = (word1, word2) => {
  let charFreqWord1 = charFreqHM(word1);

  for (let char of word2) {
    if (!(char in charFreqWord1)) {
      console.log("False");
      return false;
    }

    charFreqWord1[char] -= 1;
  }

  for (let key in charFreqWord1) {
    if (charFreqWord1[key] !== 0) {
      console.log("False");
      return false;
    }
  }

  console.log("True");
  return true;
};

anagrams("restful", "fluster"); // -> true
anagrams("cats", "tocs"); // -> false
anagrams("monkeyswrite", "newyorktimes"); // -> true
anagrams("elbow", "below"); // -> true

// JAVASCRIPT - O(2n+m) T, O(n) S

/* Approach (JavaScript)
- Create a helper function that takes a word string as an input,
processes the word and generates a HashMap of the frequency 
of each char/letter of the word.
- First word is processed and respective HashMaps is generated.
Loop over second word and check if first word HashMap firstly has
each char as a key, if it doesn't - early return false, if it does
subtract 1 from freq count.
- Lastly loop over first word HashMap and see if frequencies for chars
are equal to 0. If yes, return true.
*/

File: algorithms/hashing/anagrams/Anagrams.java
import java.util.HashMap;

public class Anagrams {
    
    private static boolean anagramsMethod(String word1, String word2) {
        HashMap<Character, Integer> word1freqHash = freqMap(word1);
        HashMap<Character, Integer> word2freqHash = freqMap(word2);

        System.out.println("Are the words anagrams? The answer is... " + word1freqHash.equals(word2freqHash));
        return word1freqHash.equals(word2freqHash);
    }

    private static HashMap<Character, Integer> freqMap(String word) {
        HashMap<Character, Integer> charFreq = new HashMap<>();

        for (char letter : word.toCharArray()) {
            if (charFreq.containsKey(letter)) {
                charFreq.put(letter, charFreq.get(letter) + 1);
            } else {
                charFreq.put(letter, 0);
            }
        }

        return charFreq;
    }

    public static void main(String[] args) {
        anagramsMethod("restful", "fluster"); // -> true
        anagramsMethod("cats", "tocs"); // -> false
        anagramsMethod("monkeyswrite", "newyorktimes"); // -> true
        anagramsMethod("elbow", "below"); // -> true
    }
}

// JAVA- O(n+m) T, O(n+m) S

/*
 * Approach
 * - Create a helper function that can be used to process a given
 * String and would generate a HashMap of the freq of each letter char.
 * - The two words that are inputted are "processed" via the
 * helper function and the resulting HashMaps of each word are compared.
 * - If they are equal, which means they have the exact same letters,
 * then return true.
 */


/*
 * Learning Keys
 * 
 * - .equals()
 */
